flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/metadata/MetadataV3Serializer.java

```diff
@@ -115,4 +120,31 @@ protected OperatorState deserializeOperatorState(DataInputStream dis) throws IOE
 
 		return operatorState;
 	}
+
+	@VisibleForTesting
+	@Override
+	public void serializeResultSubpartitionStateHandle(ResultSubpartitionStateHandle handle, DataOutputStream dos) throws IOException {
+		channelStateHandleSerializer.serialize(handle, dos);
+	}
+
+	@VisibleForTesting
+	@Override
+	public ResultSubpartitionStateHandle deserializeResultSubpartitionStateHandle(DataInputStream dis) throws IOException {
+		final boolean hasResultSubpartitionStateHandle = dis.readInt() != 0;
```

Comment:
are you sure that `dis.readInt() != 0` is good enough check to maintain backward compatibility?

Do we have test coverage for that?

Reply:
Backward compatibility is maintained on a snapshot level (by having version number at the beginning of it). This check is for cases when there is no state (same approach as for other parts of the state).

Reply:
Ok. 

Do we have proper test coverage for backward compatibility?

Reply:
Just to clarify:
- V3 serializer was added in this release (1.11); initially, it was added for FLIP-27 operator coordinators state
- the versioning mechanism was added several years ago

So any test that reads a snapshot from file covers this because these files were written with serializer V2 or older. There are a lot of such tests.